hazelcast/src/main/java/com/hazelcast/internal/dynamicconfig/ClusterWideConfigurationService.java

```diff
@@ -0,0 +1,531 @@
+
+    private final ConfigPatternMatcher configPatternMatcher;
+
+    @SuppressWarnings("unchecked")
+    private final Map<?, ? extends IdentifiedDataSerializable>[] allConfigurations = new Map[]{
+            mapConfigs,
+            multiMapConfigs,
+            cardinalityEstimatorConfigs,
+            ringbufferConfigs,
+            lockConfigs,
+            listConfigs,
+            setConfigs,
+            replicatedMapConfigs,
+            topicConfigs,
+            executorConfigs,
+            durableExecutorConfigs,
+            scheduledExecutorConfigs,
+            semaphoreConfigs,
+            queueConfigs,
+            reliableTopicConfigs,
+            cacheSimpleConfigs,
+            cacheEventJournalConfigs,
+            mapEventJournalConfigs,
+    };
+
+    private volatile Version version;
+
+    public ClusterWideConfigurationService(NodeEngine nodeEngine, DynamicConfigListener dynamicConfigListener) {
+        this.nodeEngine = nodeEngine;
+        this.listener = dynamicConfigListener;
+        this.configPatternMatcher = nodeEngine.getConfig().getConfigPatternMatcher();
+    }
+
+    @Override
+    public Operation prepareReplicationOperation(PartitionReplicationEvent event) {
+        if (version.isLessOrEqual(V3_8)) {
+            return null;
+        }
+        IdentifiedDataSerializable[] allConfigurations = getAllConfigurations();
+        if (allConfigurations.length == 0) {
+            return null;
+        }
+        return new DynamicConfigReplicationOperation(allConfigurations, true);
+    }
+
+    private IdentifiedDataSerializable[] getAllConfigurations() {
+        List<IdentifiedDataSerializable> all = new ArrayList<IdentifiedDataSerializable>();
+        for (Map<?, ? extends IdentifiedDataSerializable> entry : allConfigurations) {
+            Collection<? extends IdentifiedDataSerializable> values = entry.values();
+            all.addAll(values);
+        }
+        return all.toArray(new IdentifiedDataSerializable[0]);
+    }
+
+    @Override
+    public void beforeMigration(PartitionMigrationEvent event) {
+        //no-op
+    }
+
+    @Override
+    public void commitMigration(PartitionMigrationEvent event) {
+        //no-op
+    }
+
+    @Override
+    public void rollbackMigration(PartitionMigrationEvent event) {
+        //no-op
+    }
+
+    @Override
+    public void onClusterVersionChange(Version newVersion) {
+        version = newVersion;
+    }
+
+    @Override
+    public void init(NodeEngine nodeEngine, Properties properties) {
+        listener.onServiceInitialized(this);
+    }
+
+    @Override
+    public void reset() {
+        for (Map<?, ?> entry : allConfigurations) {
+            entry.clear();
+        }
+    }
+
+    @Override
+    public void shutdown(boolean terminate) {
+        //no-op
+    }
+
+    @Override
+    public void broadcastConfig(IdentifiedDataSerializable config) {
+        // we create an defensive copy as local operation execution might use a fast-path
```

Comment:
Shouldn't this method check cluster version for rolling upgrade compat? Or I am missing the place where version check is done...